### ISSUE TITLE
refactor: using bitwise operations to save gas

### ISSUE THREAD

**Issue by 0xneves** [blockful-io/swaplace-contracts#104]

**Comment by blackbeard002** [comment]
Hey @0xneves I'm wondering how the date value will look like in 'expiry' inside the struct. Let's say the expiry date is 25th december 2023. so does the value in 'expiry' be '25122023'?

**Comment by blackbeard002** [comment]
Hey @0xneves can I take this up?

**Comment by 0xneves** [comment]
It's been a while since I slipped this, sorry for keeping you waiting Captain!
How are you feeling about taking up the bitwise journey? I can assist you in this as well.

**Comment by blackbeard002** [comment]
The changes are done in the contracts. It's the just testing part that's getting delayed. Will make a pr pretty soon 😀

**Comment by 0xneves** [comment]
Hey @blackbeard002 I'm reviewing the PR and I think this will help both code legibility and the tests:

This is from the [ccip swaplace](https://github.com/blockful-io/ccip/blob/main/contracts/CCIP.sol) we are working.
```
  function packData(
    address allowed,
    uint64 destinationChainSelector,
    uint32 expiration
  ) public pure returns (uint256) {
    return
      uint160(allowed) |
      (uint256(destinationChainSelector) << 160) |
      (uint256(expiration) << (160 + 64));
  }

  function parseData(
    uint256 validationData
  ) public pure returns (address, uint64, uint32) {
    return (
      address(uint160(validationData)),
      uint64(validationData >> 160),
      uint32(validationData >> 224)
    );
  }
```

By adding the following inside the SwapFactory.sol, we can aid in creating swaps and also with legibility.

After adapting to uint160 and uint96 it should look something like this (not tested):
```
  function packData(
    address allowed,
    uint96 expiration
  ) public pure returns (uint256) {
    return
      uint160(allowed) |
      (uint256(expiration) << (160));
  }

  function parseData(
    uint256 validationData
  ) public pure returns (address, uint96) {
    return (
      address(uint160(validationData)),
      uint96(validationData >> 160),
    );
  }
```

Function are public to aid external calls when building swaps.

**Comment by blackbeard002** [comment]
@0xneves Yeah, this is much neater work. I'll redo and test according to this method. 

However, there are instances when we only need 'expiry' and not 'allowed.' In such cases, how about parsing 'expiry' the way I've currently implemented? For instances where we do need both 'expiry' and 'allowed,' we can use the parseData() method you suggested. This way, we keep the code neat and reduce unnecessary gas costs incurred by type casting and shifting operations when 'allowed' is not needed.

**Comment by 0xneves** [comment]
> @0xneves Yeah, this is much neater work. I'll redo and test according to this method.
> 
> However, there are instances when we only need 'expiry' and not 'allowed.' In such cases, how about parsing 'expiry' the way I've currently implemented? For instances where we do need both 'expiry' and 'allowed,' we can use the parseData() method you suggested. This way, we keep the code neat and reduce unnecessary gas costs incurred by type casting and shifting operations when 'allowed' is not needed.
> 
> These are the current implementation of how I'm parsing,
> 
> address allowed = address(uint160(swap.config >> 96)); uint256 expiry = swap.config & ((1 << 96) - 1);
> 
> These are tested. And thus I was trying avoid the gas cost of type castings and shifting operation.

Yes, agreed in parts, I think the parser is needed when building swaps using the Factory or when the function uses both variables parsed, for instance, createSwap uses both allowed and expiry, thus using `(address allowed, uint256 expiry) = parseData(swap.config)` will not have much of a gas difference and will preserve overall legibility. I think the only use case for that would be in cancelSwap where only expiry is needed.

I think you will like to help when we build Swaplace for Ethereum, because we will probably use assembly for everything, even the following:
```
address allowed = address(uint160(swap.config >> 96));
uint256 expiry = swap.config & ((1 << 96) - 1);
```
will be straight assembly. 

We are almost at the end of optimizing Swaplace for the minimal launch for overall EVMs. Then we will split into the Ethereum version and the ccip version for overall EVM. Later on, the overall ccip for Ethereum.

So my thoughts on this, is that we should use the parse and the pack functions for clearance and for the finalization of this issue, then we can dig into optimization as another issue.

What do you say @blackbeard002 ?

**Comment by blackbeard002** [comment]
@0xneves yeah sounds good. Another issue for optimization should work. For now I'll just implement the parseData() method. I'll redo and push the changes in the contracts. Once you take a look at it and approve I'll start with the unit tests.

**Comment by blackbeard002** [comment]
@0xneves Made the changes in the contracts. parseData() and packData() implemented. Shall I go ahead with the unit tests?

**Comment by blackbeard002** [comment]
Hey @0xneves, I have re-written the unit tests as well. PR #165 closes this current issue.
